pkg/resources/schemas/template.go

```diff
@@ -42,82 +45,108 @@ type Store struct {
+
+	schemaChangeSignal, err := s.schemaChangeNotify(notifyCtx)
+	if err != nil {
+		notifyCancel()
+		return nil, fmt.Errorf("failed to start schema change notifications: %w", err)
+	}
+
+	userChangeSignal := s.userChangeNotify(notifyCtx, user)
 
 	go func() {
+		defer notifyCancel()
 		defer wg.Done()
-		schemas, err := s.sf.Schemas(user)
-		if err != nil {
-			logrus.Errorf("failed to generate schemas for notify user %v: %v", user, err)
-			return
-		}
-		for range s.userChangeNotify(apiOp.Context(), user) {
+
+		// For each change notification send schema updates onto the result channel.
+		for {
+			select {
+			case _, ok := <-schemaChangeSignal:
+				if !ok {
+					return
+				}
+			case _, ok := <-userChangeSignal:
+				if !ok {
+					return
+				}
+			}
 			schemas = s.sendSchemas(result, apiOp, user, schemas)
 		}
 	}()
 
 	return result, nil
 }
 
+// sendSchemas will send APIEvents onto the provided result channel based on detected changes in the schemas for the provided users.
 func (s *Store) sendSchemas(result chan types.APIEvent, apiOp *types.APIRequest, user user.Info, oldSchemas *types.APISchemas) *types.APISchemas {
+	// get the current schemas for a user
 	schemas, err := s.sf.Schemas(user)
 	if err != nil {
 		logrus.Errorf("failed to get schemas for %v: %v", user, err)
 		return oldSchemas
 	}
 
 	inNewSchemas := map[string]bool{}
-	for _, apiObject := range schemastore.FilterSchemas(apiOp, schemas.Schemas).Objects {
+
+	// Convert the schemas for the given user to a flat list of APIObjects.
+	apiObjects := schemastore.FilterSchemas(apiOp, schemas.Schemas).Objects
+	for i := range apiObjects {
+		apiObject := apiObjects[i]
 		inNewSchemas[apiObject.ID] = true
 		eventName := types.ChangeAPIEvent
+
+		// Check to see if the schema represented by the current APIObject exist in the oldSchemas.
 		if oldSchema := oldSchemas.LookupSchema(apiObject.ID); oldSchema == nil {
 			eventName = types.CreateAPIEvent
 		} else {
 			newSchemaCopy := apiObject.Object.(*types.APISchema).Schema.DeepCopy()
 			oldSchemaCopy := oldSchema.Schema.DeepCopy()
 			newSchemaCopy.Mapper = nil
 			oldSchemaCopy.Mapper = nil
+
+			// APIObjects are intentionally stripped of access information. Thus we will remove the field when comparing changes.
+			delete(oldSchemaCopy.Attributes, "access")
 			if equality.Semantic.DeepEqual(newSchemaCopy, oldSchemaCopy) {
 				continue
 			}
 		}
+
+		// Send the new or modified schema as an APIObject on the APIEvent channel.
 		result <- types.APIEvent{
 			Name:         eventName,
 			ResourceType: "schema",
 			Object:       apiObject,
 		}
 	}
 
-	for _, oldSchema := range schemastore.FilterSchemas(apiOp, oldSchemas.Schemas).Objects {
+	// Identify all of the oldSchema APIObjects that have been removed and send Remove APIEvents.
+	oldSchemaObjs := schemastore.FilterSchemas(apiOp, oldSchemas.Schemas).Objects
+	for i := range oldSchemaObjs {
+		oldSchema := oldSchemaObjs[i]
```

Comment:
same comment

Reply:
Same as above to prevent unneeded copies of the object.